cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -369,11 +369,11 @@ func StartControllers(s *options.CMServer, kubeconfig *restclient.Config, rootCl
 	// Find the list of namespaced resources via discovery that the namespace controller must manage
 	namespaceKubeClient := client("namespace-controller")
 	namespaceClientPool := dynamic.NewClientPool(restclient.AddUserAgent(kubeconfig, "namespace-controller"), restMapper, dynamic.LegacyAPIPathResolverFunc)
-	groupVersionResources, err := namespaceKubeClient.Discovery().ServerPreferredNamespacedResources()
+	deletableGroupVersionResources, err := namespaceKubeClient.Discovery().ServerPreferredNamespacedResourcesWithVerbs("delete")
```

Comment:
@liggitt I have changed the `ServerPreferredResources` signature to return `APIResourceLists`. So we have all the info we can get from discovery. There are also two convenience functions `ExtractGroupVersionResources(WithVerbs)` to get the old behavior.

I have adapter the two controllers. Ptal at the namespace controller especially. I have not remove the `opCache` yet, but pre-filled it with the verbs from discovery. Probably one could rip out a whole lot of logic for graceful fallback to list+delete in the controller.